axelor-account/src/main/java/com/axelor/apps/account/service/invoice/InvoiceLineServiceImpl.java

```diff
@@ -635,30 +635,34 @@ public List<InvoiceLine> updateLinesAfterFiscalPositionChange(Invoice invoice)
 
         FiscalPosition fiscalPosition = invoice.getFiscalPosition();
         boolean isPurchase = InvoiceToolService.isPurchase(invoice);
-        TaxLine taxLine = this.getTaxLine(invoice, invoiceLine, isPurchase);
-        invoiceLine.setTaxLine(taxLine);
-        invoiceLine.setTaxRate(taxLine.getValue());
-        invoiceLine.setTaxCode(taxLine.getTax().getCode());
-
-        TaxEquiv taxEquiv =
-            accountManagementAccountService.getProductTaxEquiv(
-                invoiceLine.getProduct(), invoice.getCompany(), fiscalPosition, isPurchase);
-        invoiceLine.setTaxEquiv(taxEquiv);
-
-        Account account =
-            accountManagementAccountService.getProductAccount(
-                invoiceLine.getProduct(),
-                invoice.getCompany(),
-                fiscalPosition,
-                isPurchase,
-                invoiceLine.getFixedAssets());
-        invoiceLine.setAccount(account);
-        invoiceLine.setInTaxTotal(
-            invoiceLine
-                .getExTaxTotal()
-                .multiply(invoiceLine.getTaxRate())
-                .setScale(2, RoundingMode.HALF_UP));
-        invoiceLine.setCompanyInTaxTotal(invoiceLine.getInTaxTotal());
+
+        // Lines are not updated if product is not filled
+        if (invoiceLine.getProduct() != null) {
+          TaxLine taxLine = this.getTaxLine(invoice, invoiceLine, isPurchase);
+          invoiceLine.setTaxLine(taxLine);
+          invoiceLine.setTaxRate(taxLine.getValue());
+          invoiceLine.setTaxCode(taxLine.getTax().getCode());
+
+          TaxEquiv taxEquiv =
+              accountManagementAccountService.getProductTaxEquiv(
+                  invoiceLine.getProduct(), invoice.getCompany(), fiscalPosition, isPurchase);
+          invoiceLine.setTaxEquiv(taxEquiv);
+
+          Account account =
+              accountManagementAccountService.getProductAccount(
+                  invoiceLine.getProduct(),
+                  invoice.getCompany(),
+                  fiscalPosition,
+                  isPurchase,
+                  invoiceLine.getFixedAssets());
+          invoiceLine.setAccount(account);
+          invoiceLine.setInTaxTotal(
+              invoiceLine
+                  .getExTaxTotal()
+                  .multiply(invoiceLine.getTaxRate())
+                  .setScale(2, RoundingMode.HALF_UP));
+          invoiceLine.setCompanyInTaxTotal(invoiceLine.getInTaxTotal());
+        }
```

Comment:
```suggestion
        if (invoiceLine.getProduct() == null) {
          continue;
        }
        
        TaxLine taxLine = this.getTaxLine(invoice, invoiceLine, isPurchase);
        invoiceLine.setTaxLine(taxLine);
        invoiceLine.setTaxRate(taxLine.getValue());
        invoiceLine.setTaxCode(taxLine.getTax().getCode());

        TaxEquiv taxEquiv =
            accountManagementAccountService.getProductTaxEquiv(
                invoiceLine.getProduct(), invoice.getCompany(), fiscalPosition, isPurchase);
        invoiceLine.setTaxEquiv(taxEquiv);

        Account account =
            accountManagementAccountService.getProductAccount(
                invoiceLine.getProduct(),
                invoice.getCompany(),
                fiscalPosition,
                isPurchase,
                invoiceLine.getFixedAssets());
        invoiceLine.setAccount(account);
        invoiceLine.setInTaxTotal(
            invoiceLine
                .getExTaxTotal()
                .multiply(invoiceLine.getTaxRate())
                .setScale(2, RoundingMode.HALF_UP));
        invoiceLine.setCompanyInTaxTotal(invoiceLine.getInTaxTotal());
```

https://medium.com/@scadge/if-statements-design-guard-clauses-might-be-all-you-need-67219a1a981a

Reply:
Also, the `else` on right above this block (https://github.com/axelor/axelor-open-suite/pull/8957/files#diff-cfb3cff182ff5b0f15a324f499386c119301aa55c8a9a2e989980923b76f288eR633) seems useless. Shouldn't we remove it?

Reply:
Please apply these 2 refactoring suggestions on the other files you modified, they are built similarly.